### ISSUE TITLE
Unable to validate freed Object

### ISSUE THREAD

**Issue by ashtonmeuser** [godotengine/godot-cpp#1587]
### Godot version

4.3

### godot-cpp version

4.3 (e298f430b55bdbd9cf3eae9e041354c26d2e9c79)

### System information

macOS Sonoma 14.6.1

### Issue description

With the removal of `UtilityFunctions::is_instance_valid()` (https://github.com/godotengine/godot-cpp/pull/1513), it is impossible to validate an Object passed to a GDExtension method as a Variant.

Spawned from [this comment](https://github.com/godotengine/godot-cpp/pull/1513#issuecomment-2347275208). Using `ObjectDB::get_instance()` as suggested [here](https://github.com/godotengine/godot-cpp/issues/1390#issuecomment-1937570699) and [here](https://github.com/godotengine/godot-cpp/pull/1513#issue-2384393056) isn't a perfect replacement. Starting with a Variant that represents a freed Object, we're unable to get an object ID in order to use ObjectDB::get_instance(). After ensuring a Variant is of type Object, an operator Object*() call causes a crash.

Previously, `UtilityFunctions::is_instance_valid()` could be used to validate an object. It's possible to use `variant.stringify() == "<Freed Object>"` as a hacky workaround although obviously not ideal.

### Steps to reproduce

Add the following method to a GDExtension class.

```c++
// Must validate supplied object
godot_error MyLib::dummy(const Variant v) {
  // Side effects for other Variant types would necessitate accepting a Variant rather than Object*
  if (v.get_type() != Variant::OBJECT) return ERR_CANT_CREATE;
  // if (v.stringify() == "<Freed Object>") return ERR_CANT_CREATE; // Hacky solution
  auto o = v.operator Object*(); // Crashes here if object was previously freed
  if (o == NULL) return ERR_CANT_CREATE;
  auto id = o->get_instance_id();
  auto x = ObjectDB::get_instance(id);
  if (x == NULL) return ERR_CANT_CREATE;
  return OK;
}
```

Use the GDExtension function via GDScript as follows.

```gdscript
var o = Object.new()
var lib = MyLib.new()
assert(lib.dummy(o) == OK) # Expect object to be valid
o.free()
assert(lib.dummy(o) != OK) # Expect object to be invalid
```

### Minimal reproduction project

N/A

**Comment by ashtonmeuser** [comment]
@dsnopek Agreed that something like `Variant::get_validated_object()` would be useful. Also agree that while `is_instance_valid()` is a bit of a footgun, it's better than lacking an equivalent tool in v4.3.

A more realistic example than I provided above (and the case that revealed this issue to me) is a method that accepts a higher-order data structure e.g. Dictionary that predicate a Variant but is expected to contain an Object.

**Comment by dsnopek** [comment]
I just posted PR https://github.com/godotengine/godot-cpp/pull/1592 which reverts the removal of `is_instance_valid()`, targeting Godot 4.3 and earlier.

And also PR https://github.com/godotengine/godot-cpp/pull/1591 which adds what we need to support this use-case without `is_instance_valid()` that we can use in Godot 4.4+

**Comment by aaronfranke** [comment]
I also ran into this. I tried to use `Variant::get_validated_object` to match core, and it was not available.

**Comment by dsnopek** [comment]
This should be fixed now since https://github.com/godotengine/godot-cpp/pull/1591 and https://github.com/godotengine/godot-cpp/pull/1592 have been merged
